promql/metric.go

```diff
@@ -0,0 +1,207 @@
+	// no list was found, look for a mandatory = qual
+	name := equalQuals["name"].GetStringValue()
+	if name == "" {
+		return nil, fmt.Errorf("name parameter required")
+	}
+	return []string{name}, nil
+}
+
+func listMetric(ctx context.Context, d *plugin.QueryData, h *plugin.HydrateData) (interface{}, error) {
+
+	log := plugin.Logger(ctx)
+
+	// metric name(s)
+	aliases, err := getAliasesFromQuery(ctx, d)
+	if err != nil {
+		return nil, err
+	}
+
+	// date range
+	from, to, err := getTimerange(ctx, d)
+	if err != nil {
+		return nil, err
+	}
+	step := (to.Sub(from)/1000 + time.Second/2).Round(time.Second)
+	r := v1api.Range{
+		Start: from,
+		End:   to,
+		Step:  step,
+	}
+
+	client, err := getPromClient(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, alias := range aliases {
+
+		queryTemplate, found := getAliasDef(alias)
+		if !found {
+			queryTemplate = alias + "{ {{.filter}} }"
+		}
+		tmpl, err := template.New("t").Parse(queryTemplate)
+		if err != nil {
+			return nil, err
+		}
+
+		var buf bytes.Buffer
+		if err := tmpl.Execute(&buf, nil); err != nil {
+			return nil, fmt.Errorf("could not render query: %s", err)
+		}
+
+		result, warnings, err := client.QueryRange(ctx, buf.String(), r)
+		if err != nil {
+			log.Trace("query", buf.String())
+			return nil, fmt.Errorf("could not execute query: %s", err)
+		}
+		for _, w := range warnings {
+			log.Trace("promql.alias", "query_warning", w)
+		}
+		var rows int
+		for _, i := range result.(model.Matrix) {
+			i.Metric["__name__"] = model.LabelValue(alias)
+			for _, v := range i.Values {
+				row := model.Sample{
+					Metric:    i.Metric,
+					Timestamp: v.Timestamp,
+					Value:     v.Value,
+				}
+				d.StreamListItem(ctx, row)
+				rows += 1
+			}
+		}
+		log.Trace("got rows", rows)
+	}
+
+	return nil, nil
+}
+
+func getTimerange(ctx context.Context, d *plugin.QueryData) (from, to time.Time, err error) {
+	quals := d.Quals
+	if quals["timestamp"] != nil {
+		for _, q := range quals["timestamp"].Quals {
+			switch q.Operator {
+			case ">":
+				from = q.Value.GetTimestampValue().AsTime()
+			case ">=":
+				from = q.Value.GetTimestampValue().AsTime()
+			case "<":
+				to = q.Value.GetTimestampValue().AsTime()
+			case "<=":
+				to = q.Value.GetTimestampValue().AsTime()
+			}
+		}
+	}
+	// default to 1 hour if no range is specified
+	if from.IsZero() && to.IsZero() {
+		to = time.Now()
+		from = time.Now().Add(-time.Hour)
+	}
+	if from.IsZero() {
+		return from, to, fmt.Errorf("starting timestamp must be specified: %s", from)
+	}
+	if to.IsZero() {
+		return from, to, fmt.Errorf("ending timestamp must be specified: %s", to)
+	}
+	return from, to, nil
+}
+
+func getMetric(ctx context.Context, d *plugin.QueryData, h *plugin.HydrateData) (interface{}, error) {
+	return nil, fmt.Errorf("unimplemented get metric")
+}
+
+func getLabelsWithoutName(_ context.Context, d *transform.TransformData) (interface{}, error) {
+	ls := d.Value.(model.Metric)
+	m := make(map[string]string)
```

Comment:
Potential optimization
```suggestion
	m := make(map[string]string, len(ls)-1)
```